### PR TITLE
fix: use up-to-date chainId/accounts when querying EIP1193-derived wallets

### PR DESCRIPTION
**Pull request by zzmp**
Serially requests `accounts` and then `chainId`, so that if a user changes chains whilst connecting, the new chain (after connection / `accounts`) will be used. Practically, this prevents web3-react from using incorrect initial state when connecting to certain wallets, in certain cases where the user selects a non-active chain. This is resolved by waiting for `accounts` to settle before requesting `chainId`.

---

For `chainId` and `accounts`, the values were fetched using a `Promise.all`. In some wallets, `chainId` resolves immediately, and then the user is prompted to change chains. In these cases, `chainId` will be outdated by the time `accounts` resolves:
1. `eth_chainId` resolves to "stale" chain
2. user is prompted to connect with a chain selector
3. user selects a different "active" chain
4. `chainChanged` is emitted
5. `eth_accounts` resolves, resolving the `Promise.all` with an out-of-date `chainId`

NB: WalletConnect did not exhibit this bug, but I updated it to keep it in sync with other connectors' idioms. It likely already had this "fix" in it, based on existing comments.

### CODE REVIEW

File: packages/metamask/src/index.ts [hunk 1]
```diff
@@ -96,11 +108,11 @@ export class MetaMask extends Connector {
...
 
-    return Promise.all([
-      this.provider.request({ method: 'eth_chainId' }) as Promise<string>,
-      this.provider.request({ method: 'eth_accounts' }) as Promise<string[]>,
-    ])
-      .then(([chainId, accounts]) => {
+    return Promise.resolve()
+      .then(() => {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const { chainId, accounts }= this.provider!
```
Comment (NoahZinsmeister) [inline packages/metamask/src/index.ts:114]: nit, no space after `}`

File: packages/metamask/src/index.ts [hunk 2]
```diff
@@ -133,10 +145,10 @@ export class MetaMask extends Connector {
       .then(async () => {
         if (!this.provider) throw new NoMetaMaskError()
 
-        return Promise.all([
-          this.provider.request({ method: 'eth_chainId' }) as Promise<string>,
-          this.provider.request({ method: 'eth_requestAccounts' }) as Promise<string[]>,
```
Comment (NoahZinsmeister) [inline packages/metamask/src/index.ts:138]: haven't tested but i think this is a problem - we can't remove `eth_requestAccounts`, otherwise an unconnected user will never be able to trigger a connection popup

**Review by NoahZinsmeister — COMMENTED** [review]
wow this is really subtle, nice catch. i have a concern about the metamask change, see comment

the non-null assertions felt a little clunky but i went through them all and i didn't see anything problematic; it's maybe slightly worse because we're now relying on the coinbase wallet/metamask providers to set properties correctly, but doesn't really bother me

**Review by NoahZinsmeister — APPROVED** [review]
lgtm, nice catch

File: packages/coinbase-wallet/src/index.ts [hunk 3]
```diff
@@ -85,17 +85,14 @@ export class CoinbaseWallet extends Connector {
...
-        if (!accounts.length) throw new Error('No accounts returned')
-        this.actions.update({ chainId: parseChainId(chainId), accounts })
-      })
+      if (!this.provider || !this.connected) throw new Error('No existing connection')
+
+      // Wallets may resolve eth_chainId and hang on eth_accounts pending user interaction, which may include changing
+      // chains; they should be requested serially, with accounts first, so that the chainId can settle.
+      const accounts = await this.provider.request<string[]>({ method: 'eth_accounts' }) 
+      const chainId = await this.provider.request<string>({ method: 'eth_chainId' }) 
+      if (!accounts.length) throw new Error('No accounts returned')
```
Comment (grabbou) [inline packages/coinbase-wallet/src/index.ts:94]: nit: can we do this before requesting chainId?

File: packages/coinbase-wallet/src/index.ts [hunk 4]
```diff
@@ -144,38 +139,36 @@ export class CoinbaseWallet extends Connector {
...
-              })
-            }
-
-            throw error
-          })
-      })
+      // Wallets may resolve eth_chainId and hang on eth_accounts pending user interaction, which may include changing
+      // chains; they should be requested serially, with accounts first, so that the chainId can settle.
+      const accounts = await this.provider.request<string[]>({ method: 'eth_requestAccounts' })
+      const chainId = await this.provider.request<string>({ method: 'eth_chainId' })
```
Comment (grabbou) [inline packages/coinbase-wallet/src/index.ts:147]: nit: probably out of scope for this PR, but I would consider adding to a todo list: potential candidate for abstracting away? Looks like a common code path here and few lines earlier.

File: packages/eip1193/src/index.ts [hunk 5]
```diff
@@ -42,39 +42,30 @@ export class EIP1193 extends Connector {
...
-      this.provider.request({ method: 'eth_accounts' }) as Promise<string[]>,
-    ])
-      .then(([chainId, accounts]) => {
-        this.actions.update({ chainId: parseChainId(chainId), accounts })
-      })
-      .catch((error) => {
+    try {
+      // Wallets may resolve eth_chainId and hang on eth_accounts pending user interaction, which may include changing
+      // chains; they should be requested serially, with accounts first, so that the chainId can settle.
+      const accounts = await (eager ? this.provider.request({ method: 'eth_accounts' }) : this.provider
```
Comment (grabbou) [inline packages/eip1193/src/index.ts:51]: Question about this `eager` flag. I don't see it being used before? We were running `eth_accounts` in both instances. Is this change intentional?

Comment (grabbou) [reply packages/eip1193/src/index.ts:51]: Ok, I see this has to do with `eager` that can be `true` or `false`, depending on the call-site. Is there a better way to describe its intent? Is it something like connect if already authorised vs explicitly authorise? In that case, why fallback to `eth_accounts` when `eth_requestAccounts` failed? This is probably a bit out of scope for this PR, but a good opportunity to expand my `web3-react` knowledge a bit!

File: packages/metamask/src/index.ts [hunk 6]
```diff
@@ -133,42 +137,40 @@ export class MetaMask extends Connector {
...
-                  params: [{ ...desiredChainIdOrChainParameters, chainId: desiredChainIdHex }],
-                })
-              }
-
-              throw error
-            })
-            .then(() => this.activate(desiredChainId))
+        // Wallets may resolve eth_chainId and hang on eth_accounts pending user interaction, which may include changing
+        // chains; they should be requested serially, with accounts first, so that the chainId can settle.
+        const accounts = await this.provider.request({ method: 'eth_requestAccounts' }) as string[]
```
Comment (grabbou) [inline packages/metamask/src/index.ts:142]: follow-up to my previous question: why no fallback to `eth_accounts` here?

File: packages/walletconnect/src/index.ts [hunk 7]
```diff
@@ -125,15 +125,13 @@ export class WalletConnect extends Connector {
...
 
-      // for walletconnect, we always use sequential instead of parallel fetches because otherwise
-      // chainId defaults to 1 even if the connecting wallet isn't on mainnet
-      const accounts = await this.provider?.request<string[]>({ method: 'eth_accounts' })
+      // Wallets may resolve eth_chainId and hang on eth_accounts pending user interaction, which may include changing
+      // chains; they should be requested serially, with accounts first, so that the chainId can settle.
+      const accounts = await this.provider.request<string[]>({ method: 'eth_accounts' })
+      const chainId = await this.provider.request<string>({ method: 'eth_chainId' })
       if (!accounts.length) throw new Error('No accounts returned')
-      const chainId = await this.provider
```
Comment (grabbou) [inline packages/walletconnect/src/index.ts:132]: why this change? I actually commented in other place asking whether we shouldn't actually return early instead of letting `chainId` request to start when no accounts returned.

File: packages/walletconnect/src/index.ts [hunk 8]
```diff
@@ -125,15 +125,13 @@ export class WalletConnect extends Connector {
...
-        .request<string | number>({ method: 'eth_chainId' })
-        .then((chainId) => parseChainId(chainId))
 
-      this.actions.update({ chainId, accounts })
+      this.actions.update({ chainId: parseChainId(chainId), accounts })
```
Comment (grabbou) [inline packages/walletconnect/src/index.ts:134]: nit: maybe `chainId` should already be parsed? `parseChainId(await ....)`, otherwise I think this line is a bit confusing: `chainId: parseChainId(chainId)` vs `this.actions.update({ chainId })`

Comment (grabbou) [reply packages/walletconnect/src/index.ts:134]: If considered a worthy nit, this most likely applies to other files too.

**Review by grabbou — APPROVED** [review]
Looks good to me! Left some overall nits / comments, but nothing blocking

File: packages/coinbase-wallet/src/index.ts (hunk 4 above)
Comment (zzmp) [reply packages/coinbase-wallet/src/index.ts:147]: I think that web3-react *is* the abstraction but, more importantly, I don't want to make this a large change because it's meant to be a bug fix.

File: packages/eip1193/src/index.ts (hunk 5 above)
Comment (zzmp) [reply packages/eip1193/src/index.ts:51]: I believe this is done because EIP1193 does not necessarily support eth_requestAccounts, so a fallback is necessary. You don't see this in other connectors because they are not explicitly EIP1193 connectors.

The codepaths for connectEagerly and activate were identical except for the usage of eth_accounts vs eth_requestAccounts, so I added connect as a shared method to express that with less code duplication.

I've refactored to `activateAccounts(requestAccounts: () => Promise<string[]>)`.

File: packages/metamask/src/index.ts (hunk 6 above)
Comment (zzmp) [reply packages/metamask/src/index.ts:142]: MetaMask is known to implement this method, whereas a pure EIP1193 may not.

File: packages/walletconnect/src/index.ts (hunk 8 above)
Comment (zzmp) [reply packages/walletconnect/src/index.ts:134]: fwiw, I agree with you here, but I don't want to change it because it expands the scope of this PR - I've omitted it to de-risk this bug fix.
